File: debate-engine/src/types/debate.ts
export interface DebateState {
  id: string;
  question: string;
  context?: string;
  config: DebateConfig;
  rounds: DebateRound[];
  status: "running" | "converged" | "max_rounds" | "error";
  convergenceReason?: string;
  finalReport?: DecisionReport;
  metadata: {
    startedAt: string;
    completedAt?: string;
    totalLLMCalls: number;
    totalTokensUsed: number;
  };
}

export interface DebateConfig {
  maxRounds: number;
  convergenceThreshold: number;
  model: string;
  temperature: number;
  language: "zh" | "en";
}

export interface DebateRound {
  roundNumber: number;
  phases: {
    advocate: AgentResponse;
    critic: AgentResponse;
    factCheck: FactCheckResponse;
    moderator: ModeratorResponse;
  };
}

export interface AgentResponse {
  agentRole: string;
  arguments: Argument[];
  rebuttals: Rebuttal[];
  concessions: string[];
  confidenceShift: number;
}

export interface Argument {
  id: string;
  claim: string;
  reasoning: string;
  evidence?: string;
  status: "active" | "rebutted" | "conceded" | "modified";
}

export interface Rebuttal {
  targetArgumentId: string;
  counterClaim: string;
  reasoning: string;
}

export interface FactCheckResponse {
  checks: FactCheck[];
  overallAssessment: string;
}

export interface FactCheck {
  targetArgumentId: string;
  verdict: "valid" | "flawed" | "needs_context" | "unverifiable";
  explanation: string;
  correction?: string;
}

export interface ModeratorResponse {
  roundSummary: string;
  keyDivergences: string[];
  convergenceScore: number;
  shouldContinue: boolean;
  guidanceForNextRound?: string;
}

export interface DecisionReport {
  question: string;
  executiveSummary: string;
  recommendation: {
    direction: string;
    confidence: "high" | "medium" | "low";
    conditions: string[];
  };
  proArguments: ScoredArgument[];
  conArguments: ScoredArgument[];
  resolvedDisagreements: string[];
  unresolvedDisagreements: string[];
  riskFactors: string[];
  nextSteps: string[];
  debateStats: {
    totalRounds: number;
    argumentsRaised: number;
    argumentsSurvived: number;
    convergenceAchieved: boolean;
  };
}

export interface ScoredArgument {
  claim: string;
  strength: number;
  survivedChallenges: number;
  modifications: string[];
}

export interface AgentInput {
  question: string;
  context?: string;
  config: DebateConfig;
  history: DebateRound[];
  currentRound: number;
  moderatorGuidance?: string;
  currentRoundAdvocate?: AgentResponse;
  currentRoundCritic?: AgentResponse;
  currentRoundFactCheck?: FactCheckResponse;
}

File: maverickj/schemas/debate.py
from datetime import datetime
from enum import Enum
from typing import Optional, Any

from pydantic import BaseModel, Field

from maverickj.schemas.agents import AgentResponse, FactCheckResponse, ModeratorResponse
from maverickj.schemas.arguments import ArgumentRecord


class DebateStatus(str, Enum):
    RUNNING = "running"
    CONVERGED = "converged"
    MAX_ROUNDS = "max_rounds"
    ERROR = "error"


class DebateConfig(BaseModel):
    max_rounds: int = 5
    convergence_threshold: int = 2
    convergence_score_target: float = 0.8
    language: str = "auto"
    transcript_compression_after_round: int = 2


class DebateMetadata(BaseModel):
    started_at: datetime
    completed_at: Optional[datetime] = None
    total_llm_calls: int = 0
    total_tokens_used: int = 0
    total_cost_usd: float = 0.0


class DebateRound(BaseModel):
    round_number: int
    advocate: AgentResponse
    critic: AgentResponse
    fact_check: FactCheckResponse
    moderator: ModeratorResponse


class DebateState(BaseModel):
    """Shared state object passed between all LangGraph nodes."""
    id: str
    question: str
    context: Optional[str] = None
    config: DebateConfig
    rounds: list[DebateRound] = []
    argument_registry: dict[str, ArgumentRecord] = {}
    current_round: int = 0
    status: DebateStatus = DebateStatus.RUNNING
    convergence_reason: Optional[str] = None
    final_report: Optional[Any] = None  # DecisionReport, use Any to avoid circular import
    metadata: DebateMetadata
    # Transient fields for current round in-progress data
    current_round_advocate: Optional[AgentResponse] = None
    current_round_critic: Optional[AgentResponse] = None
    current_round_fact_check: Optional[FactCheckResponse] = None
    current_round_moderator: Optional[ModeratorResponse] = None
